File: client-sdks/sockudo-kotlin/lib/src/main/kotlin/io/sockudo/client/Push.kt
package io.sockudo.client

import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

data class PushRegistrationOptions(
    val endpoint: String,
    val headers: Map<String, String> = emptyMap(),
    val headersProvider: (() -> Map<String, String>)? = null,
)

data class PushCursorParams(
    val limit: Int? = null,
    val cursor: String? = null,
) {
    fun toQueryParams(): Map<String, Any> =
        buildMap {
            limit?.let { put("limit", it) }
            cursor?.let { put("cursor", it) }
        }
}

data class PushSubscriptionParams(
    val channel: String? = null,
    val deviceId: String? = null,
    val limit: Int? = null,
    val cursor: String? = null,
) {
    fun toQueryParams(): Map<String, Any> =
        buildMap {
            channel?.let { put("channel", it) }
            deviceId?.let { put("deviceId", it) }
            limit?.let { put("limit", it) }
            cursor?.let { put("cursor", it) }
        }
}

class SockudoPushRegistration(
    options: PushRegistrationOptions,
    private val httpClient: OkHttpClient = OkHttpClient(),
) {
    private val endpoint = options.endpoint.trimEnd('/')
    private val headers = options.headers
    private val headersProvider = options.headersProvider

    suspend fun activateDevice(device: Map<String, Any?>): Map<String, Any?> =
        requestObject("POST", "/deviceRegistrations", body = device)

    suspend fun updateDeviceRegistration(
        device: Map<String, Any?>,
        deviceIdentityToken: String,
    ): Map<String, Any?> =
        requestObject(
            "POST",
            "/deviceRegistrations",
            body = device,
            requestHeaders = mapOf("X-Sockudo-Device-Identity-Token" to deviceIdentityToken),
        )

    suspend fun listDeviceRegistrations(
        params: PushCursorParams = PushCursorParams(),
    ): Map<String, Any?> = requestObject("GET", "/deviceRegistrations", queryParams = params.toQueryParams())

    suspend fun getDeviceRegistration(deviceId: String): Map<String, Any?> =
        requestObject("GET", "/deviceRegistrations/${encodePath(deviceId)}")

    suspend fun deleteDeviceRegistration(deviceId: String) {
        requestUnit("DELETE", "/deviceRegistrations/${encodePath(deviceId)}")
    }

    suspend fun upsertChannelSubscription(subscription: Map<String, Any?>): Map<String, Any?> =
        requestObject("POST", "/channelSubscriptions", body = subscription)

    suspend fun listChannelSubscriptions(
        params: PushSubscriptionParams = PushSubscriptionParams(),
    ): Map<String, Any?> = requestObject("GET", "/channelSubscriptions", queryParams = params.toQueryParams())

    suspend fun deleteChannelSubscriptions(params: PushSubscriptionParams) {
        requestUnit("DELETE", "/channelSubscriptions", queryParams = params.toQueryParams())
    }

    suspend fun publish(request: Map<String, Any?>): Map<String, Any?> =
        requestObject("POST", "/publish", body = request + ("sync" to false))

    suspend fun publishBatch(requests: List<Map<String, Any?>>): Any? =
        requestValue(
            "POST",
            "/batch/publish",
            body = requests.map { it + ("sync" to false) },
        )

    suspend fun schedulePublish(request: Map<String, Any?>): Map<String, Any?> = publish(request)

    suspend fun getPublishStatus(publishId: String): Map<String, Any?> =
        requestObject("GET", "/publish/${encodePath(publishId)}/status")

    suspend fun cancelScheduledPublish(publishId: String) {
        requestUnit("DELETE", "/scheduled/${encodePath(publishId)}")
    }

    suspend fun postDeliveryStatus(event: Map<String, Any?>): Map<String, Any?> =
        requestObject("POST", "/deliveryStatus", body = event)

    private suspend fun requestObject(
        method: String,
        path: String,
        body: Any? = null,
        queryParams: Map<String, Any> = emptyMap(),
        requestHeaders: Map<String, String> = emptyMap(),
    ): Map<String, Any?> = requestValue(method, path, body, queryParams, requestHeaders) as? Map<String, Any?> ?: emptyMap()

    private suspend fun requestUnit(
        method: String,
        path: String,
        body: Any? = null,
        queryParams: Map<String, Any> = emptyMap(),
        requestHeaders: Map<String, String> = emptyMap(),
    ) {
        requestValue(method, path, body, queryParams, requestHeaders)
    }

    private suspend fun requestValue(
        method: String,
        path: String,
        body: Any? = null,
        queryParams: Map<String, Any> = emptyMap(),
        requestHeaders: Map<String, String> = emptyMap(),
    ): Any? {
        val url =
            "${endpoint}${path}".toHttpUrlOrNull()?.newBuilder()?.apply {
                queryParams.forEach { (name, value) -> addQueryParameter(name, value.toString()) }
            }?.build()
                ?: throw SockudoException.InvalidUrl("Invalid push endpoint ${endpoint}${path}")

        val requestBody =
            if (body == null) {
                null
            } else {
                JsonSupport.encode(body).toRequestBody("application/json".toMediaType())
            }

        val request =
            Request.Builder()
                .url(url)
                .method(method, requestBody)
                .apply {
                    (headers + (headersProvider?.invoke() ?: emptyMap()) + requestHeaders).forEach { (name, value) ->
                        addHeader(name, value)
                    }
                    addHeader("Content-Type", "application/json")
                }
                .build()

        val response = httpClient.newCall(request).execute()
        response.use {
            if (!it.isSuccessful) {
                throw SockudoException.InvalidOptions(
                    "Sockudo push request failed with HTTP ${it.code}",
                )
            }
            if (it.code == 204) {
                return null
            }
            val raw = it.body?.string().orEmpty()
            if (raw.isBlank()) {
                return emptyMap<String, Any?>()
            }
            return JsonSupport.fromJsonElement(JsonSupport.decode(raw))
        }
    }

    private fun encodePath(value: String): String = java.net.URLEncoder.encode(value, Charsets.UTF_8)
}

File: client-sdks/sockudo-flutter/lib/src/push.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'support.dart';

typedef PushHeadersProvider = Map<String, String> Function();

class PushRegistrationOptions {
  const PushRegistrationOptions({
    required this.endpoint,
    this.headers = const <String, String>{},
    this.headersProvider,
  });

  final String endpoint;
  final Map<String, String> headers;
  final PushHeadersProvider? headersProvider;
}

class PushCursorParams {
  const PushCursorParams({this.limit, this.cursor});

  final int? limit;
  final String? cursor;

  Map<String, String> toQueryParameters() => <String, String>{
    'limit': ?(limit == null ? null : '$limit'),
    'cursor': ?cursor,
  };
}

class PushSubscriptionParams extends PushCursorParams {
  const PushSubscriptionParams({
    this.channel,
    this.deviceId,
    super.limit,
    super.cursor,
  });

  final String? channel;
  final String? deviceId;

  @override
  Map<String, String> toQueryParameters() => <String, String>{
    'channel': ?channel,
    'deviceId': ?deviceId,
    ...super.toQueryParameters(),
  };
}

class SockudoPushRegistration {
  SockudoPushRegistration(this.options, {http.Client? httpClient})
    : _httpClient = httpClient ?? http.Client(),
      _endpoint = options.endpoint.replaceAll(RegExp(r'/+$'), '');

  final PushRegistrationOptions options;
  final http.Client _httpClient;
  final String _endpoint;

  Future<Map<Object?, Object?>> activateDevice(Map<String, Object?> device) =>
      _requestObject('POST', '/deviceRegistrations', body: device);

  Future<Map<Object?, Object?>> updateDeviceRegistration(
    Map<String, Object?> device,
    String deviceIdentityToken,
  ) => _requestObject(
    'POST',
    '/deviceRegistrations',
    body: device,
    requestHeaders: <String, String>{
      'X-Sockudo-Device-Identity-Token': deviceIdentityToken,
    },
  );

  Future<Map<Object?, Object?>> listDeviceRegistrations([
    PushCursorParams params = const PushCursorParams(),
  ]) => _requestObject(
    'GET',
    '/deviceRegistrations',
    queryParameters: params.toQueryParameters(),
  );

  Future<Map<Object?, Object?>> getDeviceRegistration(String deviceId) =>
      _requestObject(
        'GET',
        '/deviceRegistrations/${Uri.encodeComponent(deviceId)}',
      );

  Future<void> deleteDeviceRegistration(String deviceId) => _requestVoid(
    'DELETE',
    '/deviceRegistrations/${Uri.encodeComponent(deviceId)}',
  );

  Future<Map<Object?, Object?>> upsertChannelSubscription(
    Map<String, Object?> subscription,
  ) => _requestObject('POST', '/channelSubscriptions', body: subscription);

  Future<Map<Object?, Object?>> listChannelSubscriptions([
    PushSubscriptionParams params = const PushSubscriptionParams(),
  ]) => _requestObject(
    'GET',
    '/channelSubscriptions',
    queryParameters: params.toQueryParameters(),
  );

  Future<void> deleteChannelSubscriptions(PushSubscriptionParams params) =>
      _requestVoid(
        'DELETE',
        '/channelSubscriptions',
        queryParameters: params.toQueryParameters(),
      );

  Future<Map<Object?, Object?>> publish(Map<String, Object?> request) =>
      _requestObject('POST', '/publish', body: <String, Object?>{
        ...request,
        'sync': false,
      });

  Future<Object?> publishBatch(List<Map<String, Object?>> requests) =>
      _requestValue(
        'POST',
        '/batch/publish',
        body: requests
            .map(
              (request) => <String, Object?>{...request, 'sync': false},
            )
            .toList(growable: false),
      );

  Future<Map<Object?, Object?>> schedulePublish(Map<String, Object?> request) =>
      publish(request);

  Future<Map<Object?, Object?>> getPublishStatus(String publishId) =>
      _requestObject(
        'GET',
        '/publish/${Uri.encodeComponent(publishId)}/status',
      );

  Future<void> cancelScheduledPublish(String publishId) => _requestVoid(
    'DELETE',
    '/scheduled/${Uri.encodeComponent(publishId)}',
  );

  Future<Map<Object?, Object?>> postDeliveryStatus(Map<String, Object?> event) =>
      _requestObject('POST', '/deliveryStatus', body: event);

  Future<Map<Object?, Object?>> _requestObject(
    String method,
    String path, {
    Object? body,
    Map<String, String> queryParameters = const <String, String>{},
    Map<String, String> requestHeaders = const <String, String>{},
  }) async {
    final value = await _requestValue(
      method,
      path,
      body: body,
      queryParameters: queryParameters,
      requestHeaders: requestHeaders,
    );
    return value as Map<Object?, Object?>? ?? const <Object?, Object?>{};
  }

  Future<void> _requestVoid(
    String method,
    String path, {
    Object? body,
    Map<String, String> queryParameters = const <String, String>{},
    Map<String, String> requestHeaders = const <String, String>{},
  }) async {
    await _requestValue(
      method,
      path,
      body: body,
      queryParameters: queryParameters,
      requestHeaders: requestHeaders,
    );
  }

  Future<Object?> _requestValue(
    String method,
    String path, {
    Object? body,
    Map<String, String> queryParameters = const <String, String>{},
    Map<String, String> requestHeaders = const <String, String>{},
  }) async {
    final uri = Uri.parse(
      '$_endpoint$path',
    ).replace(queryParameters: queryParameters.isEmpty ? null : queryParameters);
    final request = http.Request(method, uri)
      ..headers.addAll(<String, String>{
        'Content-Type': 'application/json',
        ...options.headers,
        ...?options.headersProvider?.call(),
        ...requestHeaders,
      });
    if (body != null) {
      request.body = jsonEncode(body);
    }
    final response = await _httpClient.send(request);
    final text = await response.stream.bytesToString();
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw SockudoException(
        'Sockudo push request failed with HTTP ${response.statusCode}',
        statusCode: response.statusCode,
      );
    }
    if (response.statusCode == 204 || text.isEmpty) {
      return const <Object?, Object?>{};
    }
    return JsonSupport.decode(text);
  }
}
